libraries/anvill_passes/src/ConvertXorToCmp.cpp

```diff
@@ -0,0 +1,198 @@
+ private:
+  static char ID;
+};
+
+char ConvertXorToCmp::ID = '\0';
+
+// Get which operand for a binary operator is an ICmpInst
+// nullptr if neither
+static llvm::ICmpInst *getComparisonOperand(llvm::BinaryOperator *op) {
+  auto rhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(1));
+  if (rhs_cmp) {
+    return rhs_cmp;
+  }
+
+  auto lhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(0));
+  if (lhs_cmp) {
+    return lhs_cmp;
+  }
+
+  return nullptr;
+}
+
+static llvm::Value *negateCmpPredicate(llvm::ICmpInst *cmp) {
+  auto pred = cmp->getPredicate();
+  llvm::IRBuilder<> ir(cmp);
+  llvm::ICmpInst::Predicate new_pred = llvm::CmpInst::getInversePredicate(pred);
+
+  // create a new compare with negated predicate
+  return ir.CreateICmp(new_pred, cmp->getOperand(0), cmp->getOperand(1));
+}
+
+bool ConvertXorToCmp::runOnFunction(llvm::Function &func) {
+  std::vector<llvm::BinaryOperator *> xors;
+
+  for (auto &inst : llvm::instructions(func)) {
+
+    // check for binary op
+    if (auto binop = llvm::dyn_cast<llvm::BinaryOperator>(&inst)) {
+
+      // binary op is a xor
+      if (binop->getOpcode() == llvm::Instruction::Xor) {
+
+        // xor has a constant as either lhs or rhs, and its a one-bit constant of size 1
+        auto lhs_c = llvm::dyn_cast<llvm::ConstantInt>(binop->getOperand(0));
+        auto lhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(binop->getOperand(0));
+
+        auto rhs_c = llvm::dyn_cast<llvm::ConstantInt>(binop->getOperand(1));
+        auto rhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(binop->getOperand(1));
+
+
+        // left side is a constant int 1 (true), right side is a cmpinst
+        if (lhs_c && lhs_c->getType()->getBitWidth() == 1 &&
+            lhs_c->isAllOnesValue() && rhs_cmp) {
+          xors.emplace_back(binop);
+        } else if (rhs_c && rhs_c->getType()->getBitWidth() == 1 &&
+            rhs_c->isAllOnesValue() && lhs_cmp) {
+          // right side is a constant int 1 (true), left side is a cmp
+          xors.emplace_back(binop);
+        }
+      }
+    }
+  }
+
+  auto changed = false;
+  int replaced_items = 0;
+
+  for (auto xori : xors) {
+
+    // find predicate from xor's operands
+    auto cmp = getComparisonOperand(xori);
+    if (!cmp) {
+      DLOG(ERROR)
+          << "Error: Xor should have a comparison operand, but we can't find it again";
+      continue;
+    }
+
+    std::vector<llvm::BranchInst *> brs_to_invert;
+    std::vector<llvm::SelectInst *> selects_to_invert;
+    bool invertible_xor = true;
+
+    //
+    for (auto &U : cmp->uses()) {
+      llvm::Instruction *inst = llvm::dyn_cast<llvm::Instruction>(U.getUser());
+
+      // use is not the existing xor
+      if (inst == xori) {
+        continue;
+      }
+
+      llvm::BranchInst *br = llvm::dyn_cast<llvm::BranchInst>(inst);
+      if (br) {
+        brs_to_invert.emplace_back(br);
+        continue;
+      }
+
+      llvm::SelectInst *si = llvm::dyn_cast<llvm::SelectInst>(inst);
+      if (si) {
```

Comment:
Here you should be more strict. That is, you should be checking if the use by the `select` is in fact the condition operand, and not a selected value.

Reply:
👌